### PR TITLE
Parse tokens from all dependencies instead of single files in the language server

### LINKED ISSUE #2318: Reorganize language server to handle full projects instead of single files
Will involve reorganizing how the `document.rs` and `session.rs` files are currently structured. We will also want to write some nice helper methods around the `TokenMap` to get all tokens for a specific file etc...

This should mostly solve #271 once implemented properly

### LINKED ISSUE #2319: Start using the typed AST tokens in the language server
At the moment we are only using non-typed tokens collected during traversal of the `ParseProgram` from `sway-core`. Most of the work for collecting typed tokens has already been done in #2121. We just need to properly handle working with full projects instead of single files first which is being tracked in #2318. 

After this is done we can start adding back LSP capabilities that require knowledge of the tokens type information.

### PR DESCRIPTION
**Pull request by JoshuaBatty**
This PR now traverses the submodules in a project and collects those tokens into the `TokenMap`. 

I've moved a lot of the logic that was previously in `document.rs` into `session.rs`. Also ended moving / simplifying a bit of code so it's easier to read.

Finally, I've changed our `TokenMap` type to use a `DashMap` instead of a `HashMap`.

closes #2318 

partially addresses #324

### CODE REVIEW

File: sway-lsp/src/core/session.rs [hunk 1]
```diff
@@ -19,16 +29,98 @@ pub type Documents = DashMap<String, TextDocument>;
...
 impl Session {
     pub fn new() -> Self {
         Session {
             documents: DashMap::new(),
             config: RwLock::new(SwayConfig::default()),
+            token_map: DashMap::new(),
+        }
+    }
+
+    /// Check if the code editor's cursor is currently over an of our collected tokens
```
Comment (mitchmindtree) [inline sway-lsp/src/core/session.rs:44]:
```suggestion
    /// Check if the code editor's cursor is currently over one of our collected tokens.
```

File: sway-lsp/src/core/session.rs [hunk 2]
```diff
@@ -19,16 +29,98 @@ pub type Documents = DashMap<String, TextDocument>;
...
+                (key.clone(), token.clone())
+            })
+            .collect()
+    }
+
+    pub fn declared_token_ident(&self, token: &TokenType) -> Option<Ident> {
+        // Look up the tokens TypeId
+        match utils::token::type_id(token) {
+            Some(type_id) => {
+                tracing::info!("type_id = {:#?}", type_id);
```
Comment (mitchmindtree) [inline sway-lsp/src/core/session.rs:102]: Is this left over from debugging?

File: sway-lsp/src/core/session.rs [hunk 3]
```diff
@@ -19,16 +29,98 @@ pub type Documents = DashMap<String, TextDocument>;
...
+
+                // Use the TypeId to look up the actual type
+                let type_info = sway_core::type_engine::look_up_type_id(type_id);
+                tracing::info!("type_info = {:#?}", type_info);
```
Comment (mitchmindtree) [inline sway-lsp/src/core/session.rs:106]: Same here

File: sway-lsp/src/server.rs [hunk 4]
```diff
@@ -70,24 +72,27 @@ fn capabilities() -> ServerCapabilities {
...
+        diagnostics: Vec<Diagnostic>,
+        token_map: &TokenMap,
+    ) {
         // If parsed_tokens_as_warnings is true, take over the normal error and warning display behavior
         // and instead show the parsed tokens as warnings.
         // This is useful for debugging the lsp parser.
         if self.config.parsed_tokens_as_warnings {
-            if let Some(document) = self.session.documents.get(uri.path()) {
-                let diagnostics = debug::generate_warnings_for_parsed_tokens(document.token_map());
+            //let diagnostics = debug::generate_warnings_for_parsed_tokens(token_map);
```
Comment (mitchmindtree) [inline sway-lsp/src/server.rs:85]: Left over from debugging?

**Review by mitchmindtree — COMMENTED** [review]
Nice one, looking much cleaner! Just a couple nits, otherwise happy to approve :heavy_check_mark:

File: sway-lsp/src/core/session.rs (hunk 3 above)
Comment (JoshuaBatty) [reply sway-lsp/src/core/session.rs:106]: Yep good spots

File: sway-lsp/src/server.rs (hunk 4 above)
Comment (JoshuaBatty) [reply sway-lsp/src/server.rs:85]: I actually had the wrong line commented out, thanks for this.

**Comment by JoshuaBatty** [comment]
> Nice one, looking much cleaner! Just a couple nits, otherwise happy to approve ✔️

Thanks, @mitchmindtree just addressed those nits, should be good to go.

**Review by mitchmindtree — APPROVED** [review]
:shipit:

**Review by otrho — APPROVED** [review]

**Pull request merged**